### PR TITLE
8311815: Incorrect exhaustivity computation

### PR DESCRIPTION
**Pull request by lahodaj**
Consider code like:
```
class Test22 {

  record Pair(I i1, I i2) {}

  sealed interface I {}

  record C() implements I {}

  record D() implements I {}

  void exhaustinvenessWithInterface(Pair pairI) {
    switch (pairI) {
      case Pair(D fst, C snd) -> {}
      case Pair(C fst, C snd) -> {}
      case Pair(C fst, I snd) -> {}
      case Pair(D fst, D snd) -> {}
    }
  }
} 
```

when evaluating exhaustivity, the outcome is wrong: if `Pair(D fst, C snd)` and `Pair(C fst, C snd)` get merged into `Pair(I _, C_)` (and the existing patterns are removed from the pattern set), the remaining patterns are `Pair(I _, C_)`, `Pair(C fst, I snd)` and `Pair(D fst, D snd)`. The existing implementation does not know how to continue from here (it would need to replace `Pair(C fst, I snd)` with `Pair(C fst, D snd)`, which would allow to unify `Pair(C fst, D snd)` and `Pair(D fst, D snd)` into `Pair(I _, D _)` and then continue, but that is tricky, as such speculation may not lead to a good result, and we would need to implement backtracking).

But, if `Pair(D fst, C snd)` and `Pair(D fst, D snd)` was merged into `Pair(D _, I _)`, it could then be merge with `Pair(C fst, I snd)` to produce `Pair(I _, I _)`.

The proposal here is aiming to allow the second path, by not removing the existing patterns when merging sealed subtypes. I.e. starting with `{Pair(D _, C _), Pair(C _, C_), Pair(C _, I _), Pair(D _, D _)}`, we would merge to `{Pair(D _, C _), Pair(C _, C_), Pair(C _, I _), Pair(D _, D _), *Pair(I _, C_)*}`, and then `{Pair(D _, C _), Pair(C _, C_), Pair(C _, I _), Pair(D _, D _), Pair(I _, C_), *Pair(D _, I _)*}`, which would then lead to `Pair(I _, I _)`, leading to `Pair` and confirmed exhaustivity.

This can be achieved easily by not removing binding in `reduceBindingPatterns`. The problem is that when we stop removing bindings, the algorithm may get into an infinite loop re-adding the same binding over and over. The solution to that is using Sets instead of Lists. To make the algorithm faster and more stable, there are two more additions: a) all record patterns covered by binding patterns are removed from the pattern set; b) `reduceNestedPatterns` attempts to produce results for all mismatching positions without returning (otherwise the algorithm may constantly only process the first mismatching position without progressing).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311815](https://bugs.openjdk.org/browse/JDK-8311815): Incorrect exhaustivity computation (**Bug** - P3)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [1f5c1735](https://git.openjdk.org/jdk/pull/14872/files/1f5c17354bbd07bcf7e7c721511b09523cf42bec)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14872/head:pull/14872` \
`$ git checkout pull/14872`

Update a local copy of the PR: \
`$ git checkout pull/14872` \
`$ git pull https://git.openjdk.org/jdk.git pull/14872/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14872`

View PR using the GUI difftool: \
`$ git pr show -t 14872`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14872.diff">https://git.openjdk.org/jdk/pull/14872.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14872#issuecomment-1634453898)

### CODE REVIEW

**Review by vicente-romero-oracle — APPROVED** [review]
looks sensible

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java [hunk 1]
```diff
@@ -3505,6 +3508,9 @@ record RecordPattern(Type recordType, int _hashCode, Type[] fullComponentTypes,
 
         public RecordPattern(Type recordType, Type[] fullComponentTypes, PatternDescription[] nested) {
             this(recordType, hashCode(-1, recordType, nested), fullComponentTypes, nested);
+            if ("test.Test.R(test.Test.C _, test.Test.R(test.Test.C _, test.Test.R(test.Test.C _, test.Test.C _, test.Test.C _), test.Test.I _), test.Test.I _)".equals(toString())) {
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java:3511]: sorry missed this, this seems like testing code that should be removed right?

Comment (lahodaj) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java:3511]: Oops, sorry, will fix.

**Pull request closed without merging**